pkg/api/controller_test.go

```diff
@@ -8145,6 +8145,439 @@ func TestRouteFailures(t *testing.T) {
+	ctlr.Config.Storage.RootDirectory = t.TempDir()
+
+	cm := test.NewControllerManager(ctlr)
+	cm.StartAndWait(port)
+
+	defer cm.StopServer()
+
+	client := resty.New()
+	client.SetBasicAuth(username, password)
+
+	img := CreateRandomImage()
+
+	repoNames := []string{
+		"alpine1", "alpine2", "alpine3",
+		"alpine4", "alpine5", "alpine6",
+		"alpine7", "alpine8", "alpine9",
+	}
+
+	for _, repo := range repoNames {
+		err := UploadImageWithBasicAuth(img, baseURL, repo, "0.1", username, password)
+		if err != nil {
+			panic(err)
+		}
+	}
+
+	conf.HTTP.AccessControl.Repositories = config.Repositories{
+		"alpine[13579]": readPolicyGroup,
+		"alpine[2468]":  config.PolicyGroup{},
+	}
+
+	// Note empty strings signify the query parameter is not set
+	// There are separate tests for passing the empty string as query parameter
+	testCases := []struct {
+		testCaseName  string
+		pageSize      string
+		last          string
+		expectedRepos []string
+	}{
+		{
+			testCaseName:  "no parameters",
+			pageSize:      "",
+			last:          "",
+			expectedRepos: []string{"alpine1", "alpine3", "alpine5", "alpine7", "alpine9"},
+		},
+		{
+			testCaseName:  "first 3",
+			pageSize:      "3",
+			last:          "",
+			expectedRepos: []string{"alpine1", "alpine3", "alpine5"},
+		},
+		{
+			testCaseName:  "next 2",
+			pageSize:      "3",
+			last:          "alpine5",
+			expectedRepos: []string{"alpine7", "alpine9"},
+		},
+		{
+			testCaseName:  "0",
+			pageSize:      "0",
+			last:          "",
+			expectedRepos: []string{},
+		},
+		{
+			testCaseName:  "Test the parameter 'last' without parameter 'n'",
+			pageSize:      "",
+			last:          "alpine3",
+			expectedRepos: []string{"alpine5", "alpine7", "alpine9"},
+		},
+		{
+			testCaseName:  "Test the parameter 'last' with the final repo as value",
+			pageSize:      "",
+			last:          "alpine9",
+			expectedRepos: []string{},
+		},
+	}
+
+	for _, testCase := range testCases {
+		Convey(testCase.testCaseName, t, func() {
+			testHTTPPagedRepositories(t, client, baseURL, testCase.testCaseName, testCase.pageSize,
+				testCase.last, testCase.expectedRepos, repoNames[len(repoNames)-1])
+		})
+	}
+}
+
+func TestPagedRepositoriesWithSubpaths(t *testing.T) {
+	port := test.GetFreePort()
+	baseURL := test.GetBaseURL(port)
+	conf := config.New()
+	conf.HTTP.Port = port
+
+	dir := t.TempDir()
+	firstSubDir := t.TempDir()
+	secondSubDir := t.TempDir()
+
+	subPaths := make(map[string]config.StorageConfig)
+
+	subPaths["/a"] = config.StorageConfig{RootDirectory: firstSubDir}
+	subPaths["/b"] = config.StorageConfig{RootDirectory: secondSubDir}
+
+	ctlr := makeController(conf, dir)
+	ctlr.Config.Storage.SubPaths = subPaths
+	ctlr.Config.Storage.Commit = true
+
+	cm := test.NewControllerManager(ctlr)
+	cm.StartAndWait(port)
+
+	defer cm.StopServer()
+
+	rthdlr := api.NewRouteHandler(ctlr)
+
+	img := CreateRandomImage()
+
+	repoNames := []string{
+		"alpine1", "alpine2", "alpine3",
```

Comment:
Can we mix them so the results from the default store and substores are not in alphabetical order?

Let's say substores with paths /a and /c, and images "alpine1", "alpine2", "alpine3", "a/alpine4", "a/alpine5", "a/alpine6", "b/alpine7", "b/alpine8", "b/alpine9", "c/alpine10", "c/alpine11", "c/alpine12", "d/alpine10", "d/alpine11", "d/alpine12". Also some entries not having a folder in path.